daffodil-core/src/main/scala/org/apache/daffodil/dpath/Expression.scala

```diff
@@ -1870,7 +1870,32 @@ case class FunctionCallExpression(functionQNameString: String, expressions: List
       case (RefQName(_, "unsignedByte", XSD), args) =>
         XSConverterExpr(functionQNameString, functionQName, args, NodeInfo.UnsignedByte)
 
-      case _ => SDE("Unsupported function: %s", functionQName)
+      case (_: RefQName, args) => {
+        val namespace = functionQName.namespace.toString()
+        val fName = functionQName.local
+
+        val udfCallingInfo = UserDefinedFunctionService.lookupUDFCallingInfo(namespace, fName)
+
+        if (udfCallingInfo.isEmpty) {
+          SDE(s"Error calling $namespace:$fName")
+        } else {
+          /*
+           * Attempting assignnment by tuple matching results in a scala existentials error in SBT
+           * so we do it this way instead
+           */
+          val udfci = udfCallingInfo.get
+          val udf = udfci._1
+          val evaluateMethod = udfci._2
+          val paramTypes = udfci._3
+          val retType = udfci._4
+
+          val evaluateParamTypes: List[NodeInfo.Kind] = paramTypes.map { c => NodeInfo.fromClass(c) }.toList
+          val evaluateReturnType = NodeInfo.fromClass(retType)
```

Comment:
We dont, so I'm going to move this bit of code into UDFService and validate there. I like idea of returning an Option instead of throwing an error. And then the caller can process the None however it likes, thoughts?

Reply:
Sounds good to me.

Reply:
I don't think this framework is called in that manner. This is the expression compiler. It expects things to type check and succeed, or it inserts conversions, or it fails with an SDE at compile time.

Reply:
Hmm almost all the validations of the UDFs are in UDFService though. I would think validating if the param types/return types of the UDF are supported during loading (and ignoring it if it's unsupported), rather than loading it and then failing in the expression compiler would be a better option?

Reply:
Disagree. The UDF service should publish what the required arg types are, and should publish what the result type will be. The Expression compiler should use that information exactly the way it does for every other built in function and operator, to type check the expression. 

UDFs should check UDF-specific things and leave type-checking to where it already lives.

Reply:
Even if a UDF uses a type that Daffodil doesn't and can't ever know about? For exampe:
``` java
class MyCustomType(...)

class MyCustomUDF {
String evaluate(MyCustomType t) = { t.getSomething.toString }
}
```
This to me seems like an invalid UDF that should be handled consistently with how we handle other invalid UDF's (wich currently means we just ignore it).

Reply:
This is my concern as well. 

As far as I know the Expression compiler does type checking of its types of the values (i.e the args or the return) and compares NodeInfo.Kind to NodeInfo.Kind. We are attempting to convert the type of the UDF params i.e in the example above MyCustomType to NodeInfo.Kind, since type checking doesn't do Class[_] to NodeInfo.Kind. If we fail to do the conversion (which is what the proposed solution is referencing), we won't even get to the type checking of the param/return value types the expression compiler does.

Reply:
Ah. You are correct. I had the wrong context here. We're trying to determine the NodeInfo.Kind type from the java arg type, and error out if there is no obvious direct mapping. That's a UDF-specific issue we could check. 

Withdraw my concerns.